### PR TITLE
MSFT contributions June 2023

### PR DESCRIPTION
**Pull request by C-Core**

### CODE REVIEW

File: Code/Engine/Foundation/Communication/Event.h [hunk 1]
```diff
@@ -140,6 +140,9 @@ class ezEventBase
   /// \brief Removes all registered event handlers.
   void Clear();
 
+  /// \brief Returns true, if not event handlers are registered.
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Communication/Event.h]: typo: "if no ..."

File: Code/BuildSystem/CMake/ezUtils.cmake [hunk 2]
```diff
@@ -188,9 +196,11 @@ function(ez_set_common_target_definitions TARGET_NAME)
 	target_compile_definitions(${TARGET_NAME} PRIVATE BUILDSYSTEM_SDKVERSION_MINOR="${EZ_CMAKE_SDKVERSION_MINOR}")
 	target_compile_definitions(${TARGET_NAME} PRIVATE BUILDSYSTEM_SDKVERSION_PATCH="${EZ_CMAKE_SDKVERSION_PATCH}")
 
+	set(ORIGINAL_BUILD_TYPE "$<IF:$<STREQUAL:${EZ_CMAKE_GENERATOR_CONFIGURATION},${EZ_BUILDTYPENAME_DEBUG}>,DEBUG,$<IF:$<STREQUAL:${EZ_CMAKE_GENERATOR_CONFIGURATION},${EZ_BUILDTYPENAME_DEV}>,DEV,SHIPPING>>")
+
 	# set the BUILDSYSTEM_BUILDTYPE definition
-	target_compile_definitions(${TARGET_NAME} PRIVATE BUILDSYSTEM_BUILDTYPE="${EZ_CMAKE_GENERATOR_CONFIGURATION}")
-	target_compile_definitions(${TARGET_NAME} PRIVATE BUILDSYSTEM_BUILDTYPE_${EZ_CMAKE_GENERATOR_CONFIGURATION})
+	target_compile_definitions(${TARGET_NAME} PRIVATE "BUILDSYSTEM_BUILDTYPE=\"${ORIGINAL_BUILD_TYPE}\"")
+	target_compile_definitions(${TARGET_NAME} PUBLIC "BUILDSYSTEM_BUILDTYPE_${ORIGINAL_BUILD_TYPE}")
```
Comment (jankrassnigg) [inline Code/BuildSystem/CMake/ezUtils.cmake:203]: Why was this change necessary ?

File: Code/Engine/Foundation/Configuration/CVar.h [hunk 3]
```diff
@@ -137,15 +148,35 @@ class EZ_FOUNDATION_DLL ezCVar : public ezEnumerable<ezCVar>
   ///
   ///
   /// \sa LoadCVarsFromCommandLine()
-  static void LoadCVarsFromFile(bool bOnlyNewOnes = true, bool bSetAsCurrentValue = true); // [tested]
+  static void LoadCVarsFromFile(bool bOnlyNewOnes = true, bool bSetAsCurrentValue = true, ezDynamicArray<ezCVar*>* pOutCVars = nullptr); // [tested]
+
+  /// \brief Loads all CVars from the given file. Does not account for any plug-in specific files are save flags.
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Configuration/CVar.h]: "specific files are save flags" ? Typo ?

File: Code/Engine/Foundation/Configuration/CVar.h [hunk 4]
```diff
@@ -137,15 +148,35 @@ class EZ_FOUNDATION_DLL ezCVar : public ezEnumerable<ezCVar>
...
+  /// Otherwise their 'Current' value will always stay unchanged and the value from disk will only be
+  /// stored in the 'Restart' value.
+  /// Independent on the parameter settings, all CVar changes during loading will always trigger change events.
+  ///
+  /// \sa LoadCVarsFromCommandLine()
+  /// \sa LoadCVarsFromFile()
+  static void LoadCVarsFromFile(ezStringView path, bool bOnlyNewOnes = true, bool bSetAsCurrentValue = true, ezDynamicArray<ezCVar*>* pOutCVars = nullptr);
 
   /// \brief Similar to LoadCVarsFromFile() but tries to get the CVar values from the command line
+  /// The CVars are loaded into the global system and thus automatically available everywhere after this call.
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Configuration/CVar.h:178]: add an empty comment line here, otherwise this is all the \brief comment

File: Code/Engine/Foundation/Configuration/CVar.h [hunk 5]
```diff
@@ -137,15 +148,35 @@ class EZ_FOUNDATION_DLL ezCVar : public ezEnumerable<ezCVar>
...
+  /// All other CVars will stay unchanged.
+  /// If \a bSetAsCurrentValue is true, variables that are flagged as 'RequiresRestart', will be set
+  /// to the restart value immediately ('SetToRestartValue' is called on them).
+  /// Otherwise their 'Current' value will always stay unchanged and the value from disk will only be
+  /// stored in the 'Restart' value.
+  /// Independent on the parameter settings, all CVar changes during loading will always trigger change events.
+  ///
+  /// \sa LoadCVarsFromCommandLine()
+  /// \sa LoadCVarsFromFile()
+  static void LoadCVarsFromFile(ezStringView path, bool bOnlyNewOnes = true, bool bSetAsCurrentValue = true, ezDynamicArray<ezCVar*>* pOutCVars = nullptr);
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Configuration/CVar.h]: "sPath", otherwise clang-tidy complains

File: Code/Engine/Foundation/Configuration/CVar.h [hunk 6]
```diff
@@ -137,15 +148,35 @@ class EZ_FOUNDATION_DLL ezCVar : public ezEnumerable<ezCVar>
...
+  /// All other CVars will stay unchanged.
+  /// If \a bSetAsCurrentValue is true, variables that are flagged as 'RequiresRestart', will be set
+  /// to the restart value immediately ('SetToRestartValue' is called on them).
+  /// Otherwise their 'Current' value will always stay unchanged and the value from disk will only be
+  /// stored in the 'Restart' value.
+  /// Independent on the parameter settings, all CVar changes during loading will always trigger change events.
+  ///
+  /// \sa LoadCVarsFromCommandLine()
+  /// \sa LoadCVarsFromFile()
+  static void LoadCVarsFromFile(ezStringView sPath, bool bOnlyNewOnes = true, bool bSetAsCurrentValue = true, ezDynamicArray<ezCVar*>* pOutCVars = nullptr);
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Configuration/CVar.h]: unit test ?

File: Code/Engine/Foundation/Configuration/CVar.h [hunk 7]
```diff
@@ -120,13 +120,24 @@ class EZ_FOUNDATION_DLL ezCVar : public ezEnumerable<ezCVar>
   /// So it might not be necessary to call this function manually at shutdown.
   static void SaveCVars(); // [tested]
 
+  /// \brief Stores all CVar values into the given file.
+  ///
+  /// This function works without setting a storage folder.
+  ///
+  /// \sa LoadCVarsFromFile()
+  static void SaveCVarsToFile(ezStringView sPath);
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Configuration/CVar.h]: unit test ?

File: Code/Engine/Foundation/Containers/Bitfield.h [hunk 8]
```diff
@@ -109,6 +109,9 @@ class ezStaticBitfield
   /// \brief Clears the given bit to 0.
   void ClearBit(ezUInt32 uiBit); // [tested]
 
+  /// \brief Sets the given bit to 1 or 0 dependen on the given value.
+  void SetBitValue(ezUInt32 uiBit, bool bValue);
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Containers/Bitfield.h]: unit test

**Review by jankrassnigg — CHANGES_REQUESTED** [review]

File: Code/BuildSystem/CMake/ezUtils.cmake (hunk 2 above)
Comment (Ingrater) [reply Code/BuildSystem/CMake/ezUtils.cmake:203]: We have a `#ifdef BUILDSYSTEM_BUILDTYPE_Shipping` in UserConfig.h
If you do not use Debug|Dev|Shipping but the CMake default Debug|RelWithDebInfo|Release instead, then the this `#ifdef` would never trigger.

Furthermore because various `#ifdef BUILDSYSTEM_BUILDTYPE_XXX` statements are visible from public headers, the visibility has to be changed from PRIVATE to PUBLIC otherwise they are only defined when building Foundation and not compiling something that consumes foundation.

File: Code/Engine/Foundation/Containers/Bitfield.h [hunk 9]
```diff
@@ -109,6 +112,9 @@ class ezStaticBitfield
   /// \brief Clears the given bit to 0.
   void ClearBit(ezUInt32 uiBit); // [tested]
 
+  /// \brief Sets the given bit to 1 or 0 depending on the given value.
+  void SetBitValue(ezUInt32 uiBit, bool bValue); // [tested]
+
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Containers/Bitfield.h:117]: duplicate code ??

File: Code/Engine/Foundation/Containers/Implementation/Bitfield_inl.h [hunk 10]
```diff
@@ -373,6 +386,19 @@ EZ_ALWAYS_INLINE void ezStaticBitfield<T>::ClearBit(ezUInt32 uiBit)
...
+  if (bValue)
+  {
+    SetBit(uiBit);
+  }
+  else
+  {
+    ClearBit(uiBit);
+  }
+}
+
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Containers/Implementation/Bitfield_inl.h:401]: duplicate code ?

**Review by jankrassnigg — CHANGES_REQUESTED** [review]

File: Code/Engine/Foundation/Containers/Bitfield.h (hunk 9 above)
Comment (C-Core) [reply Code/Engine/Foundation/Containers/Bitfield.h:117]: As discussed offline there are two versions of bitfield, ezBitfield and ezStaticBitfield both in the same file. So yes the code is duplicated but in different classes.

File: Code/Engine/Foundation/Configuration/CVar.h [hunk 11]
```diff
@@ -120,7 +120,7 @@ class EZ_FOUNDATION_DLL ezCVar : public ezEnumerable<ezCVar>
   /// So it might not be necessary to call this function manually at shutdown.
   static void SaveCVars(); // [tested]
 
-  /// \brief Stores all CVar values into the given file.
+  /// \brief Stores all CVar values into the given file, ignoring the save flag ezCVarFlags::Save.
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Configuration/CVar.h]: The fact that the Save flag is completely ignored by this function is a bit unexpected and easy to miss. It would be great if this was an option to the Save function instead, so that one can choose whether all CVars get saved or only the ones with the flag.

**Review by jankrassnigg — APPROVED** [review]
Approved with suggestions.

File: Code/Engine/Foundation/Configuration/CVar.h (hunk 11 above)
Comment (C-Core) [reply Code/Engine/Foundation/Configuration/CVar.h]: Added an option

**Review by jankrassnigg — APPROVED** [review]
